staging/src/k8s.io/client-go/rest/factory.go

```diff
@@ -0,0 +1,175 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rest
+
+import (
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+type RESTClientFactory struct {
+	// base is the root URL for all invocations of the client
+	base *url.URL
+	// rateLimiter is shared among all requests created by this client unless specifically
+	// overridden.
+	rateLimiter flowcontrol.RateLimiter
+	// creates BackoffManager that is passed to requests.
+	createBackoffMgr func() BackoffManager
+	// Set specific behavior of the client.  If not set http.DefaultClient will be used.
+	Client *http.Client
+}
+
+func RESTClientFactoryFor(config *Config) (*RESTClientFactory, error) {
+	// Base URL
+	base, _, err := defaultServerUrlFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	if !strings.HasSuffix(base.Path, "/") {
+		base.Path += "/"
+	}
+	base.RawQuery = ""
+	base.Fragment = ""
+
+	// Transport
+	if config.UserAgent == "" {
+		config.UserAgent = DefaultKubernetesUserAgent()
+	}
+
+	transport, err := TransportFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var httpClient *http.Client
+	if transport != http.DefaultTransport {
+		httpClient = &http.Client{Transport: transport}
+		if config.Timeout > 0 {
+			httpClient.Timeout = config.Timeout
+		}
+	}
+
+	// RateLimiter
+	rateLimiter := config.RateLimiter
+	if rateLimiter == nil {
+		qps := config.QPS
+		if config.QPS == 0.0 {
+			qps = DefaultQPS
+		}
+		burst := config.Burst
+		if config.Burst == 0 {
+			burst = DefaultBurst
+		}
+		if qps > 0 {
+			rateLimiter = flowcontrol.NewTokenBucketRateLimiter(qps, burst)
+		}
+	}
+
+	return &RESTClientFactory{
+		base:             base,
+		rateLimiter:      rateLimiter,
+		createBackoffMgr: readExpBackoffConfig,
+		Client:           httpClient,
+	}, err
+}
+
+func RESTClientFactoryFromClient(c RESTClient) *RESTClientFactory {
+	return &RESTClientFactory{
+		base:             c.base,
+		rateLimiter:      c.GetRateLimiter(),
+		createBackoffMgr: readExpBackoffConfig,
+		Client:           c.Client,
+	}
+}
+
+func (r RESTClientFactory) NewFor(config ClientContentConfig) *RESTClient {
+	// shallow copy
+	c := config
+	if len(c.ContentType) == 0 {
+		config.ContentType = "application/json"
+	}
+
+	// TODO: handle the case were the gv is empty?
+	c.Negotiator = runtime.NewClientNegotiator(scheme.Codecs.WithoutConversion(), c.GroupVersion)
+	versionedAPIPath := versionedAPIPathFromGV(c.GroupVersion)
+
+	return &RESTClient{
+		// shared
+		base:        r.base,
+		rateLimiter: r.rateLimiter,
+		Client:      r.Client,
+		// per client values
+		versionedAPIPath: versionedAPIPath,
+		content:          c,
+	}
+}
+
+func (r RESTClientFactory) NewForGV(group, version string) *RESTClient {
```

Comment:
I couldn't figure out how to pass the gv directly from the generator so I took this brute force approach for the sake of prototyping it